### PR TITLE
fix: Add Channels to Tenants

### PR DESCRIPTION
**Pull request by filipecabaco**
## What kind of change does this PR introduce?

Adds the concept of Channels into the Tenant database

Also adds some helper methods to handle SQL queries and results in a more "Elixir" way

### CODE REVIEW

File: lib/realtime/channels.ex [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+  def get_channel_by_name(db_conection, name) do
+    query = from c in Channel, where: c.name == ^name
+    {query, args} = Repo.to_sql(:all, query)
+
+    with res <- Postgrex.query!(db_conection, query, args),
+         [channel] <- Repo.pg_result_to_struct(res, Channel) do
+      channel
+    else
+      [] -> nil
+      _ -> raise "Multiple channels with the same name"
```
Comment (chasers) [inline lib/realtime/channels.ex:36]: I assume we will have a constraint for this too in the db?

File: lib/realtime/repo.ex [hunk 2]
```diff
@@ -15,4 +15,48 @@ defmodule Realtime.Repo do
       Supervisor.stop(repo)
     end
   end
+
+  @doc """
+  Converts a Postgrex.Result into a given struct
+  """
+  @spec pg_result_to_struct(Postgrex.Result.t(), module()) :: [struct()]
+  def pg_result_to_struct(%Postgrex.Result{rows: rows, columns: columns}, struct) do
+    Enum.map(rows, fn row ->
```
Comment (chasers) [inline lib/realtime/repo.ex:24]: Is this how Ecto does it?

File: lib/realtime/tenants/repo/migrations/20231018144023_create_channels.ex [hunk 3]
```diff
@@ -0,0 +1,14 @@
+defmodule Realtime.Tenants.Migrations.CreateChannels do
+  @moduledoc false
+
+  use Ecto.Migration
+
+  def change do
+    create table(:channels, prefix: "realtime") do
```
Comment (chasers) [inline lib/realtime/tenants/repo/migrations/20231018144023_create_channels.ex:7]: Need to check but I think with self-hosted we have the "_realtime" schema too. Might need to make this dynamic.

File: lib/realtime/tenants/repo/migrations/20231018144023_create_channels.ex [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+      add(:name, :string, null: false)
+      timestamps()
+    end
+
+    create unique_index(:channels, [:name], prefix: "realtime")
```
Comment (chasers) [inline lib/realtime/tenants/repo/migrations/20231018144023_create_channels.ex:12]: Ah yep.

**Review by chasers — APPROVED** [review]
LGTM

File: lib/realtime/api/channel.ex [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+  schema "channels" do
+    field(:name, :string)
+    timestamps()
+  end
+
+  def changeset(channel, attrs) do
+    channel
+    |> cast(attrs, [:name, :inserted_at, :updated_at])
+    |> put_timestamp(:updated_at)
+    |> maybe_put_timestamp(:inserted_at)
```
Comment (abc3) [inline lib/realtime/api/channel.ex:20]: why is that necessary?

File: lib/realtime/tenants/repo/migrations/20231018144023_create_channels.ex (hunk 3 above)
Comment (chasers) [reply lib/realtime/tenants/repo/migrations/20231018144023_create_channels.ex:7]: Ah yeah[ our metadata db uses "_realtime"](https://github.com/supabase/realtime/blob/5c7bdf802592c97f3846585673ffcd16c3579618/README.md?plain=1#L67) but we create it with self-hosted anyways.

File: lib/realtime/channels.ex [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+    :ok
+  end
+
+  @doc """
+  Fetches a channel by name from the tenant database using a given DBConnection
+  """
+  @spec get_channel_by_name(DBConnection.t(), String.t()) :: Channel.t() | nil
+  def get_channel_by_name(db_conection, name) do
+    query = from c in Channel, where: c.name == ^name
+    {query, args} = Repo.to_sql(:all, query)
```
Comment (abc3) [inline lib/realtime/channels.ex:29]: Why is it necessary to write a query in a DSL and then convert it into standard SQL?

Comment (abc3) [reply lib/realtime/channels.ex:29]: That is the same `select * from realtime.channels where name = $1?`

File: lib/realtime/channels.ex (hunk 1 above)
Comment (filipecabaco) [reply lib/realtime/channels.ex:36]: yes 👍 that is set with an unique index

File: lib/realtime/channels.ex (hunk 6 above)
Comment (chasers) [reply lib/realtime/channels.ex:29]: I think it's cool 😎

File: lib/realtime/repo.ex (hunk 2 above)
Comment (filipecabaco) [reply lib/realtime/repo.ex:24]: it's a bit tricky to find but I think so with this spot in the code
https://github.com/elixir-ecto/ecto/blob/8fac7068ed28cbb63c9844bb4cfba09fb48b9a09/lib/ecto/repo/schema.ex#L595

just needs to do this with a ton of code in the middle 😅

File: lib/realtime/tenants/repo/migrations/20231018144023_create_channels.ex (hunk 3 above)
Comment (filipecabaco) [reply lib/realtime/tenants/repo/migrations/20231018144023_create_channels.ex:7]: will change 👍

File: lib/realtime/api/channel.ex (hunk 5 above)
Comment (filipecabaco) [reply lib/realtime/api/channel.ex:20]: Since we're not using Ecto but Postgrex directly this isn't done automatically as Ecto is the one handling timestamps

File: lib/realtime/repo.ex (hunk 2 above)
Comment (filipecabaco) [reply lib/realtime/repo.ex:24]: will check in the elixir slack where does this happen to see if I can reuse it

File: lib/realtime/channels.ex (hunk 6 above)
Comment (filipecabaco) [reply lib/realtime/channels.ex:29]: Using the DSL ends up being less error prone and avoids plain sql in the code which improves readability.

Comment (abc3) [reply lib/realtime/channels.ex:29]: Sure, but we will need to call a bunch of functions and perform several lookups from ETS, which seems too much to me for a simple SQL select

File: lib/realtime/api/channel.ex (hunk 5 above)
Comment (abc3) [reply lib/realtime/api/channel.ex:20]: Can the channel records be edited in the future? Also, the `realtime.subscription` table uses the `created_at` column to indicate the creation time. So it would be nice to keep consistency

Comment (abc3) [reply lib/realtime/api/channel.ex:20]: I think the `created_at` in `realtime.subscription` is filled by default

Comment (filipecabaco) [reply lib/realtime/api/channel.ex:20]: can change but this does deviate from the ecto standard (timestamps() uses inserted_at and updated_at) 

I'm unsure if we will be able to change names or any new info we add it here, for now I will assume so.

subscriptions does it with a `created_at = now()` in the insert query.

File: lib/realtime/channels.ex (hunk 6 above)
Comment (filipecabaco) [reply lib/realtime/channels.ex:29]: this is the base to then have other more complex queries. I'm also not that concerned for performance in this query at the moment since this and other information for the Authz effort will use the Contextual Cache module if we need more reads

File: lib/realtime/tenants/repo/migrations/20231018144023_create_channels.ex (hunk 3 above)
Comment (chasers) [reply lib/realtime/tenants/repo/migrations/20231018144023_create_channels.ex:7]: No no we use the `realtime` schema on tenants not `_realtime`.
